REGRESSION (268069@main): [iOS] Date picker fails to present after dismissing with Done button

whsieh opened the pull request:
#### c9ac0cb7dfabcbf5aacdfecfde49b7b9f8936683
<pre>
REGRESSION (268069@main): [iOS] Date picker fails to present after dismissing with Done button
<a href="https://bugs.webkit.org/show_bug.cgi?id=262070">https://bugs.webkit.org/show_bug.cgi?id=262070</a>

Reviewed by Aditya Keerthi.

After the changes in 268069@main, tapping a datetime input after previously focusing and dismissing
it using the Done button silently fails to show the date picker popover a second time. This is
because we don&apos;t get a call to the `-presentationControllerDidDismiss:` delegate method in this case
where we&apos;re programmatically driving the dismissal of the popover; instead, we only get a call to
the completion handler of `-dismissViewControllerAnimated:completion:`.

I had actually caught (and prevented) this bug for Catalyst, which prompted me to add the call to
`-datePickerPopoverControllerDidDismiss:` in the completion handler specifically for Catalyst, but
for some reason didn&apos;t see that it also affected iOS as well. The fix here is simply:

1.  Remove the `PLATFORM(MACCATALYST)` in the completion handler of the dismissal animation, and...
2.  Make the logic around invoking `-datePickerPopoverControllerDidDismiss:` robust in the case
    where both the popover controller delegate method and completion block are invoked (which seems
    to be the case for month and time inputs, which use `UIDatePickerStyleWheels` instead of the
    inline style).

This was also almost caught by the existing test `fast/forms/ios/show-and-dismiss-date-input.html`,
but that test dismisses the date picker by tapping elsewhere on the screen, which narrowly avoids
this bug since we get a delegate call to `-presentationControllerDidDismiss:` in that case. We add
a new layout test that works similarly to this test, but instead simulates tapping the Done button
to dismiss the date picker.

* LayoutTests/fast/forms/ios/show-and-dismiss-date-input-after-tapping-done-expected.txt: Added.
* LayoutTests/fast/forms/ios/show-and-dismiss-date-input-after-tapping-done.html: Added.

Add a new layout test to exercise this fix.

* Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
(-[WKWebView dismissFormAccessoryView]):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView dismissFormAccessoryView]):

Augment existing test-only SPI to dismiss the date picker by simulating tapping the Done button when
a date picker popover is shown.

* Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.h:
* Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.mm:
(-[WKDatePickerPopoverController dismissDatePicker]):
(-[WKDatePickerPopoverController presentInView:sourceRect:interactionBounds:completion:]):
(-[WKDatePickerPopoverController _dispatchPopoverControllerDidDismissIfNeeded]):
(-[WKDatePickerPopoverController presentationControllerDidDismiss:]):
* Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.h:
* Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
(-[WKDateTimePicker datePickerController]):
(-[WKDateTimeInputControl setTimePickerHour:minute:]):
(-[WKDateTimeInputControl dateTimePickerCalendarType]):
(-[WKDateTimeInputControl timePickerValueHour]):
(-[WKDateTimeInputControl timePickerValueMinute]):
(-[WKDateTimeInputControl dismissWithAnimation]):

Canonical link: <a href="https://commits.webkit.org/268425@main">https://commits.webkit.org/268425@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/68577b448b99254d9cfe96d85cae40a7944f14aa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19664 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20086 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20699 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21557 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18390 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19901 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23347 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20227 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19882 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19894 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17102 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22411 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17074 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17893 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/24188 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18141 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18067 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22169 "run-api-tests-without-change (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18672 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15826 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17819 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/17728 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4697 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22174 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18497 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -12027,6 +12027,13 @@ - (BOOL)_allowAnimationControls
 
+- (void)dismissFormAccessoryView
+{
+    if (auto dateInput = self.dateTimeInputControl; [dateInput dismissWithAnimation])
```
Just to confirm, `dismissWithAnimation` eventually results in `accessoryDone` right?

whsieh replied:
Indeed — `-dismissWithAnimation` exercises logic that dismisses the popover controller programmatically, which (after this fix) will result in a call to `-accessoryDone` after the animation completes.

Thanks for the review!

whsieh committed:
REGRESSION (268069@main): [iOS] Date picker fails to present after dismissing with Done button
https://bugs.webkit.org/show_bug.cgi?id=262070

Reviewed by Aditya Keerthi.

After the changes in 268069@main, tapping a datetime input after previously focusing and dismissing
it using the Done button silently fails to show the date picker popover a second time. This is
because we don't get a call to the `-presentationControllerDidDismiss:` delegate method in this case
where we're programmatically driving the dismissal of the popover; instead, we only get a call to
the completion handler of `-dismissViewControllerAnimated:completion:`.

I had actually caught (and prevented) this bug for Catalyst, which prompted me to add the call to
`-datePickerPopoverControllerDidDismiss:` in the completion handler specifically for Catalyst, but
for some reason didn't see that it also affected iOS as well. The fix here is simply:

1.  Remove the `PLATFORM(MACCATALYST)` in the completion handler of the dismissal animation, and...
2.  Make the logic around invoking `-datePickerPopoverControllerDidDismiss:` robust in the case
    where both the popover controller delegate method and completion block are invoked (which seems
    to be the case for month and time inputs, which use `UIDatePickerStyleWheels` instead of the
    inline style).

This was also almost caught by the existing test `fast/forms/ios/show-and-dismiss-date-input.html`,
but that test dismisses the date picker by tapping elsewhere on the screen, which narrowly avoids
this bug since we get a delegate call to `-presentationControllerDidDismiss:` in that case. We add
a new layout test that works similarly to this test, but instead simulates tapping the Done button
to dismiss the date picker.

* LayoutTests/fast/forms/ios/show-and-dismiss-date-input-after-tapping-done-expected.txt: Added.
* LayoutTests/fast/forms/ios/show-and-dismiss-date-input-after-tapping-done.html: Added.

Add a new layout test to exercise this fix.

* Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
(-[WKWebView dismissFormAccessoryView]):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView dismissFormAccessoryView]):

Augment existing test-only SPI to dismiss the date picker by simulating tapping the Done button when
a date picker popover is shown.

* Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.h:
* Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.mm:
(-[WKDatePickerPopoverController dismissDatePicker]):
(-[WKDatePickerPopoverController presentInView:sourceRect:interactionBounds:completion:]):
(-[WKDatePickerPopoverController _dispatchPopoverControllerDidDismissIfNeeded]):
(-[WKDatePickerPopoverController presentationControllerDidDismiss:]):
* Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.h:
* Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
(-[WKDateTimePicker datePickerController]):
(-[WKDateTimeInputControl setTimePickerHour:minute:]):
(-[WKDateTimeInputControl dateTimePickerCalendarType]):
(-[WKDateTimeInputControl timePickerValueHour]):
(-[WKDateTimeInputControl timePickerValueMinute]):
(-[WKDateTimeInputControl dismissWithAnimation]):

Canonical link: https://commits.webkit.org/268425@main
